src/mbgl/programs/programs.cpp

```diff
@@ -10,79 +10,60 @@
-      context(context_),
-      programParameters(programParameters_) {}
+Programs::Programs(const ProgramParameters& programParameters_)
+    : programParameters(programParameters_) {}
 
 Programs::~Programs() = default;
 
-void Programs::registerWith([[maybe_unused]] gfx::ShaderRegistry& registry) noexcept {}
-
-BackgroundLayerPrograms& Programs::getBackgroundLayerPrograms() noexcept {
-    if (!backgroundPrograms) {
-        backgroundPrograms = std::make_unique<BackgroundLayerPrograms>(context, programParameters);
-    }
-    return static_cast<BackgroundLayerPrograms&>(*backgroundPrograms);   
-}
-
-RasterLayerPrograms& Programs::getRasterLayerPrograms() noexcept {
-    if (!rasterPrograms) {
-        rasterPrograms = std::make_unique<RasterLayerPrograms>(context, programParameters);
-    }
-    return static_cast<RasterLayerPrograms&>(*rasterPrograms);   
-}
-
-HeatmapLayerPrograms& Programs::getHeatmapLayerPrograms() noexcept {
-    if (!heatmapPrograms) {
-        heatmapPrograms = std::make_unique<HeatmapLayerPrograms>(context, programParameters);
-    }
-    return static_cast<HeatmapLayerPrograms&>(*heatmapPrograms);   
-}
-
-HillshadeLayerPrograms& Programs::getHillshadeLayerPrograms() noexcept {
-    if (!hillshadePrograms) {
-        hillshadePrograms = std::make_unique<HillshadeLayerPrograms>(context, programParameters);
-    }
-    return static_cast<HillshadeLayerPrograms&>(*hillshadePrograms);   
-}
-
-FillLayerPrograms& Programs::getFillLayerPrograms() noexcept {
-    if (!fillPrograms) {
-        fillPrograms = std::make_unique<FillLayerPrograms>(context, programParameters);
-    }
-    return static_cast<FillLayerPrograms&>(*fillPrograms);   
-}
-
-FillExtrusionLayerPrograms& Programs::getFillExtrusionLayerPrograms() noexcept {
-    if (!fillExtrusionPrograms) {
-        fillExtrusionPrograms = std::make_unique<FillExtrusionLayerPrograms>(context, programParameters);
-    }
-    return static_cast<FillExtrusionLayerPrograms&>(*fillExtrusionPrograms);   
-}
-
-CircleLayerPrograms& Programs::getCircleLayerPrograms() noexcept {
-    if (!circlePrograms) {
-        circlePrograms = std::make_unique<CircleLayerPrograms>(context, programParameters);
-    }
-    return static_cast<CircleLayerPrograms&>(*circlePrograms);   
-}
-
-LineLayerPrograms& Programs::getLineLayerPrograms() noexcept {
-    if (!linePrograms) {
-        linePrograms = std::make_unique<LineLayerPrograms>(context, programParameters);
-    }
-    return static_cast<LineLayerPrograms&>(*linePrograms);   
+template<typename ... T>
+void registerTypes(gfx::ShaderRegistry& registry,
+    const ProgramParameters& programParameters_)
+{
+    ( [](bool expr) {
+        if (!expr) {
+            throw std::runtime_error(
+                "Failed to register " +
+                std::string(T::Name) +
+                " with shader registry!");
+        }
+    }( registry.registerShader(
+        std::make_shared<T>(programParameters_))), ... );
```

Comment:
How `...` work again? 😅

Care to add a comment what this lambda + call to registerShader is doing?

Reply:
Yeah this does need a comment - this is a fold expression over a function call. Template varargs (a parameter pack) are expanded at compile time type-wise over another expression (function call in this case). It's like a compile time loop-unroll.